### PR TITLE
[RLlib] AlgorithmConfig docs

### PR DESCRIPTION
**Pull request by maxpumperla**
We finally got around to documenting AlgoConfigs. We did so by splitting up the previously massive "Training API" doc and gave it a bit more direction. Quick breakdown:

-  AlgoConfigs now replace the "common config" section in a more structured fashion.
- We also revised the API reference section and added resp. paragraphs there.
- The code in the touched documents is now tested (was in parts quite outdated)
- We moved some of the advanced material into special user guides to make things a bit less intimidating.

Not perfect yet, but a big step forward. we can iterate on this. 

Please review starting [in the docs build here](https://ray--29796.org.readthedocs.build/en/29796/rllib/rllib-training.html)

### CODE REVIEW

File: doc/source/rllib/core-concepts.rst [hunk 1]
```diff
@@ -90,9 +90,7 @@ which implements the proximal policy optimization algorithm in RLlib.
         config = PPOConfig().environment(env="CartPole-v0").training(train_batch_size=4000)
 
         # Train via Ray Tune.
-        # Note that Ray Tune does not yet support AlgorithmConfig objects, hence
-        # we need to convert back to old-style config dicts.
-        tune.run("PPO", config=config.to_dict())
+        tune.run("PPO", config=config.)
```
Comment (sven1977) [inline doc/source/rllib/core-concepts.rst:93]: `config.` <- remove the dot?

File: doc/source/rllib/doc_code/getting_started.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+
+# __rllib-tune-config-begin__
+import ray
+from ray import air, tune
+
+
+ray.init()
+
+config = {
```
Comment (sven1977) [inline doc/source/rllib/doc_code/getting_started.py:33]: Shouldn't we also use an AlgorithmConfig object here to unconfuse the user?
e.g.
```
config = AlgorithmConfig().training(lr=tune.grid_search([0.01, 0.001, 0.0001]))
```
... should work just fine.

Comment (sven1977) [reply doc/source/rllib/doc_code/getting_started.py:33]: The last thing we want to do is to tell the user they now have to additionally learn AlgorithmConfig, but also still master the old config dicts. These should all really disappear from our codebase and all our docs.

File: doc/source/rllib/doc_code/rllib_in_60s.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# __rllib-in-60s-begin__
```
Comment (sven1977) [inline doc/source/rllib/doc_code/rllib_in_60s.py:1]: Nice! Thanks for moving this here (from rllib/examples/documentation).

File: rllib/algorithms/crr/crr.py [hunk 4]
```diff
@@ -85,32 +85,40 @@ def training(
...
 
         Args:
             weight_type: weight type to use `bin` | `exp`.
             temperature: the exponent temperature used in exp weight type.
             max_weight: the max weight limit for exp weight type.
             advantage_type: The way we reduce q values to v_t values
-            `max` | `mean` | `expectation`. `max` and `mean` work for both
-            discrete and continuous action spaces while `expectation` only
-            works for discrete action spaces.
+                `max` | `mean` | `expectation`. `max` and `mean` work for both
```
Comment (sven1977) [inline rllib/algorithms/crr/crr.py:95]: nice catch

File: doc/source/rllib/index.rst [hunk 5]
```diff
@@ -39,59 +41,50 @@ RLlib in 60 seconds
...
+    `pip install "gym[atari]" "gym[accept-rom-license]" atari_py`.
+
+This is all you need to start coding against RLlib.
+Here is an example of running a PPO Algorithm on the
+`Taxi domain <https://www.gymlibrary.dev/environments/toy_text/taxi/>`_.
+We first create a `config` for the algorithm, which sets the right environment, and
+defines all training parameters we want.
+Next, we `build` the algorithm and `train` it for a total of `5` iterations.
+A training iteration includes parallel sample collection by the environment workers, as well as loss calculation on the collected batch and a model update.
+As a last step, we `evaluate` the trained Algorithm and and render each timestep to the shell's output:
```
Comment (sven1977) [inline doc/source/rllib/index.rst:73]: `and and` <- remove one `and`

Comment (sven1977) [reply doc/source/rllib/index.rst:73]: Actually, we don't render anymore due to some gym-problems with newer gym versions.

File: doc/source/rllib/package_ref/algorithm.rst [hunk 6]
```diff
@@ -46,8 +63,16 @@ and override one or more of its methods. Those are in particular:
...
 
-Algorithm base class (ray.rllib.algorithms.algorithm.Algorithm)
----------------------------------------------------------------
+.. _rllib-algorithm-api:
+
+Interacting with an Algorithm
+-----------------------------
+
+Once you've built an :py:class:`~ray.rllib.algorithms.algorithm_config.AlgorithmConfig`
+and retrieve an :py:class:`~ray.rllib.algorithms.algorithm.Algorithm`, you can use it
```
Comment (sven1977) [inline doc/source/rllib/package_ref/algorithm.rst:72]: maybe more clear:
-> ".. and retrieve an algorithm **from it** via the `build()` method" ...

File: doc/source/rllib/package_ref/external-app.rst [hunk 7]
```diff
@@ -0,0 +1,16 @@
+External Application API
```
Comment (sven1977) [inline doc/source/rllib/package_ref/external-app.rst:1]: Nice!

File: doc/source/rllib/package_ref/external-app.rst [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+------------------------
+
+In some cases, for instance when interacting with an externally hosted simulator or
+production environment, it makes more sense to interact with RLlib as if it were an
+independently running service, rather than RLlib hosting the simulations itself.
+This is possible via RLlib's external applications interface
+`(full documentation) <rllib-env.html#external-agents-and-applications>`__.
```
Comment (sven1977) [inline doc/source/rllib/package_ref/external-app.rst:8]: Maybe add the example scripts here as well (up to you):

```
rllib/examples/serving/cartpole_[client|server].py
rllib/examples/serving/unity3d_[client|server].py
```

File: doc/source/rllib/rllib-advanced-api.rst [hunk 9]
```diff
@@ -0,0 +1,548 @@
...
+.. code-block:: python
+
+    # All of the following configs go into Algorithm.config.
+
+    # 1) Switching *off* exploration by default.
+    # Behavior: Calling `compute_action(s)` without explicitly setting its `explore`
+    # param will result in no exploration.
+    # However, explicitly calling `compute_action(s)` with `explore=True` will
+    # still(!) result in exploration (per-call overrides default).
+    "explore": False,
```
Comment (sven1977) [inline doc/source/rllib/rllib-advanced-api.rst:155]: Let's translate these into AlgorithmConfig calls as well:
e.g.
```
# 1) Switching *off* exploration by default.
# ...
# algo_config.exploration(explore=False)
```

```
# 3) Example exploration_config usages:
# ...
# algo_config.exploration(
#   explore=True,
#   exploration_config={
#   ...
#   }  
# )
```

File: doc/source/rllib/rllib-advanced-api.rst [hunk 10]
```diff
@@ -0,0 +1,548 @@
...
+RLlib will report online training rewards, however in some cases you may want to compute
+rewards with different settings (e.g., with exploration turned off, or on a specific set
+of environment configurations). You can activate evaluating policies during training (``Algorithm.train()``) by setting
+the ``evaluation_interval`` to an int value (> 0) indicating every how many ``Algorithm.train()``
+calls an "evaluation step" is run:
+
+.. code-block:: python
+
+    # Run one evaluation step on every 3rd `Algorithm.train()` call.
+    {
```
Comment (sven1977) [inline doc/source/rllib/rllib-advanced-api.rst:210]: Same here. I do realize this is a lot of translations. Maybe do this in a follow-up PR.
But we should really not mention, give-as-example, or advertise using the old config dicts anymore.

File: rllib/algorithms/algorithm_config.py [hunk 11]
```diff
@@ -36,6 +36,26 @@
 )
 from ray.tune.logger import Logger
 
+"""TODO(jungong, sven): in "offline_data" we can potentially unify all input types
```
Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:39]: Thanks for moving this here to not clutter the docs :)

File: doc/source/rllib/rllib-training.rst [hunk 12]
```diff
@@ -827,709 +184,342 @@ For older RLlib checkpoint versions (v0.1), you can still restore an algorithm v
...
+advanced usage examples:
+
+.. autoclass:: ray.rllib.algorithms.algorithm_config.AlgorithmConfig
+    :noindex:
+
+As RLlib algorithms are fairly complex, they come with many configuration options.
+To make things easier, the common properties of algorithms are naturally grouped into
+the following categories:
+
+- :ref:`Training options <rllib-config-train>`,
```
Comment (sven1977) [inline doc/source/rllib/rllib-training.rst:303]: All lower case? or with upper case?

Training options vs training options

File: doc/source/rllib/rllib-algorithms.rst [hunk 13]
```diff
@@ -95,10 +97,8 @@ Tuned examples: `CartPole-v0 <https://github.com/ray-project/ray/blob/master/rll
 
 **BC-specific configs** (see also `common configs <rllib-training.html#common-parameters>`__):
 
-.. literalinclude:: ../../../rllib/algorithms/bc/bc.py
-   :language: python
-   :start-after: __sphinx_doc_begin__
-   :end-before: __sphinx_doc_end__
+.. autoclass:: ray.rllib.algorithms.bc.bc.BCConfig
```
Comment (sven1977) [inline doc/source/rllib/rllib-algorithms.rst:100]: Nice!

File: doc/source/rllib/rllib-advanced-api.rst [hunk 14]
```diff
@@ -0,0 +1,548 @@
...
+            self.current_difficulty = task
+
+    def curriculum_fn(train_results, task_settable_env, env_ctx):
+        # Very simple curriculum function.
+        current_task = task_settable_env.get_task()
+        new_task = current_task + 1
+        return new_task
+
+    # Setup your Algorithm's config like so:
+    config = {
```
Comment (sven1977) [inline doc/source/rllib/rllib-advanced-api.rst:473]: config = AlgorithmConfig().environment(MyEnv, env_task_fn=curriculum_fn)

File: doc/source/rllib/rllib-advanced-api.rst [hunk 15]
```diff
@@ -0,0 +1,548 @@
...
+
+
+An evaluation step runs - using its own RolloutWorkers - for ``evaluation_duration`` episodes or timesteps, depending
+on the ``evaluation_duration_unit`` setting, which can take values of either "episodes" (default) or "timesteps".
+
+
+.. code-block:: python
+
+    # Every time we run an evaluation step, run it for exactly 10 episodes.
+    {
```
Comment (sven1977) [inline doc/source/rllib/rllib-advanced-api.rst:222]: same: convert to AlgorithmConfig

File: doc/source/rllib/rllib-advanced-api.rst [hunk 16]
```diff
@@ -0,0 +1,548 @@
...
+by the number of evaluation workers (configurable via ``evaluation_num_workers``), RLlib will run the remainder of episodes
+on the first n eval RolloutWorkers and leave the remaining workers idle for that time.
+
+For examples:
+
+.. code-block:: python
+
+    # Every time we run an evaluation step, run it for exactly 10 episodes, no matter, how many eval workers we have.
+    {
+        "evaluation_duration": 10,
```
Comment (sven1977) [inline doc/source/rllib/rllib-advanced-api.rst:246]: same: convert to AlgorithmConfig

File: doc/source/rllib/rllib-advanced-api.rst [hunk 17]
```diff
@@ -0,0 +1,548 @@
...
+
+.. code-block:: python
+
+    # Every time we run an evaluation step, run it for exactly 10 episodes.
+    {
+        "evaluation_duration": 10,
+        "evaluation_duration_unit": "episodes",
+    }
+    # Every time we run an evaluation step, run it for (close to) 200 timesteps.
+    {
```
Comment (sven1977) [inline doc/source/rllib/rllib-advanced-api.rst:227]: same: convert to AlgorithmConfig

File: doc/source/rllib/rllib-advanced-api.rst [hunk 18]
```diff
@@ -0,0 +1,548 @@
...
+is supported for ``evaluation_duration``. This can be used to make the evaluation step take
+roughly as long as the concurrently ongoing training step:
+
+.. code-block:: python
+
+    # Run evaluation and training at the same time via threading and make sure they roughly
+    # take the same time, such that the next `Algorithm.train()` call can execute
+    # immediately and not have to wait for a still ongoing (e.g. b/c of very long episodes)
+    # evaluation step:
+    {
```
Comment (sven1977) [inline doc/source/rllib/rllib-advanced-api.rst:287]: same: convert to AlgorithmConfig

File: doc/source/rllib/rllib-advanced-api.rst [hunk 19]
```diff
@@ -0,0 +1,548 @@
...
+Note that with or without async evaluation, all :ref:`fault tolerance settings <rllib-scaling-guide>`,
+such as ``ignore_worker_failures`` or ``recreate_failed_workers`` will be respected and applied to the failed evaluation workers.
+
+Example:
+
+.. code-block:: python
+
+    # Having an environment that occasionally blocks completely for e.g. 10min would
+    # also affect (and block) training:
+    {
```
Comment (sven1977) [inline doc/source/rllib/rllib-advanced-api.rst:335]: same: convert to AlgorithmConfig

File: doc/source/rllib/rllib-advanced-api.rst [hunk 20]
```diff
@@ -0,0 +1,548 @@
...
+
+Switch on asynchronous evaluation, meaning, we don't wait for individual
+evaluation RolloutWorkers to complete their n episode(s) (or n timesteps), but instead:
+any evaluation RolloutWorker can cover the load of another one that failed
+or is stuck in a very long lasting environment step.
+
+.. code-block:: python
+
+    {
+        # ...
```
Comment (sven1977) [inline doc/source/rllib/rllib-advanced-api.rst:359]: same: convert to AlgorithmConfig

File: doc/source/rllib/rllib-advanced-api.rst [hunk 21]
```diff
@@ -0,0 +1,548 @@
...
+
+Use the Algorithm API and update the environment between calls to ``train()``. This example shows the algorithm being run inside a Tune function.
+This is basically the same as what the built-in `env_task_fn` API described above already does under the hood, but allows you to do even more
+customizations to your training loop.
+
+.. code-block:: python
+
+    import ray
+    from ray import tune
+    from ray.rllib.algorithms.ppo import PPO
```
Comment (sven1977) [inline doc/source/rllib/rllib-advanced-api.rst:490]: PPOConfig

File: doc/source/rllib/rllib-advanced-api.rst [hunk 22]
```diff
@@ -0,0 +1,548 @@
...
+
+    num_gpus = 0
+    num_workers = 2
+
+    ray.init()
+    tune.Tuner(
+        tune.with_resources(train, resources=tune.PlacementGroupFactory(
+            [{"CPU": 1}, {"GPU": num_gpus}] + [{"CPU": 1}] * num_workers
+        ),)
+        param_space={
```
Comment (sven1977) [inline doc/source/rllib/rllib-advanced-api.rst:515]: PPOConfig().rollouts(num_rollout_workers=num_workers).resources(num_gpus=num_gpus)

File: doc/source/rllib/rllib-advanced-api.rst [hunk 23]
```diff
@@ -0,0 +1,548 @@
...
+        },
+    ).fit()
+
+You could also use RLlib's callbacks API to update the environment on new training results:
+
+.. code-block:: python
+
+    import ray
+    from ray import tune
+    from ray.rllib.agents.callbacks import DefaultCallbacks
```
Comment (sven1977) [inline doc/source/rllib/rllib-advanced-api.rst:527]: from ray.rllib.algorithms.ppo import PPOConfig

File: doc/source/rllib/rllib-advanced-api.rst [hunk 24]
```diff
@@ -0,0 +1,548 @@
...
+            algorithm.workers.foreach_worker(
+                lambda ev: ev.foreach_env(
+                    lambda env: env.set_task(task)))
+
+    ray.init()
+    tune.Tuner(
+        "PPO",
+        param_space={
+            "env": YourEnv,
+            "callbacks": MyCallbacks,
```
Comment (sven1977) [inline doc/source/rllib/rllib-advanced-api.rst:546]: PPOConfig().callbacks(MyCallbacks).environment(YourEnv)

File: doc/source/rllib/rllib-training.rst [hunk 25]
```diff
@@ -2,769 +2,139 @@
...
+    For instance, you can train the A2C algorithm on 8 workers by specifying
+    `num_workers: 8` in a JSON string passed to ``--config``:
 
-The following is a list of the common algorithm hyper-parameters:
+    .. code-block:: bash
 
-.. code-block:: python
+        rllib train --env=PongDeterministic-v4 --run=A2C --config '{"num_workers": 8}'
 
-    # === Settings for Rollout Worker processes ===
```
Comment (sven1977) [inline doc/source/rllib/rllib-training.rst:183]: OMG! It's gone. :)

**Review by sven1977 — COMMENTED** [review]
Thanks for this awesome PR @maxpumperla !
Things are falling super nicely into place now. Just a few nits and some more conversions from dict to AlgorothmConfig in the docs needed.

File: doc/source/rllib/rllib-training.rst (hunk 25 above)
Comment (maxpumperla) [reply doc/source/rllib/rllib-training.rst:183]: haha

File: doc/source/rllib/core-concepts.rst (hunk 1 above)
Comment (maxpumperla) [reply doc/source/rllib/core-concepts.rst:93]: great catch... thank you!

File: doc/source/rllib/doc_code/getting_started.py (hunk 2 above)
Comment (maxpumperla) [reply doc/source/rllib/doc_code/getting_started.py:33]: for sure, totally agree. those are simply spots I forgot to clean up

File: doc/source/rllib/index.rst (hunk 5 above)
Comment (maxpumperla) [reply doc/source/rllib/index.rst:73]: ok, fixed

File: doc/source/rllib/package_ref/algorithm.rst (hunk 6 above)
Comment (maxpumperla) [reply doc/source/rllib/package_ref/algorithm.rst:72]: yes!

File: rllib/algorithms/algorithm_config.py (hunk 11 above)
Comment (maxpumperla) [reply rllib/algorithms/algorithm_config.py:39]: yeah, I think we still have too many of those left. let's clean them one by one.

**Comment by maxpumperla** [comment]
@sven1977 thanks for your review! All that's left is to migrate the new "Advanced API guide" to algo configs. I think this document (which simply consists of topics from the old "Training API" doc that I deemed too advanced or too obscure for a first pass) should be overhauled more substantially in a follow-up PR. 

I have the feeling that some sections could either be removed or feel a bit outdated. Probably something we should discuss first!

**Review by sven1977 — APPROVED** [review]
Sounds great! Again, thanks for the PR @maxpumperla !

LGTM now.

**Review by ericl — APPROVED** [review]

**Pull request merged**